File: examples/javascript/metafield/extensions/discount-function/src/generate_cart_run.js
import {
  OrderDiscountSelectionStrategy,
  ProductDiscountSelectionStrategy,
  DiscountClass,
} from "../generated/api";

// [START discount-function.run.cart]
export function generateCartRun(input) {
  if (!input.cart.lines.length) {
    throw new Error("No cart lines found");
  }

  const { cartLinePercentage, orderPercentage, collectionIds } = parseMetafield(
    input.discount.metafield
  );

  const hasOrderDiscountClass = input.discount.discountClasses.includes(
    DiscountClass.Order
  );
  const hasProductDiscountClass = input.discount.discountClasses.includes(
    DiscountClass.Product
  );

  if (!hasOrderDiscountClass && !hasProductDiscountClass) {
    return { operations: [] };
  }

  const operations = [];
  // [START discount-function.run.cart.add-operations]
  // Add product discounts first if available and allowed
  if (hasProductDiscountClass && cartLinePercentage > 0) {
    const cartLineTargets = input.cart.lines.reduce((targets, line) => {
      // [START discount-function.run.cart.product.in_any_collection]
      if (
        "product" in line.merchandise &&
        (line.merchandise.product.inAnyCollection || collectionIds.length === 0)
      ) {
        targets.push({
          cartLine: {
            id: line.id,
          },
        });
      }
      // [END discount-function.run.cart.product.in_any_collection]
      return targets;
    }, []);

    if (cartLineTargets.length > 0) {
      operations.push({
        productDiscountsAdd: {
          candidates: [
            {
              message: `${cartLinePercentage}% OFF PRODUCT`,
              targets: cartLineTargets,
              value: {
                percentage: {
                  value: cartLinePercentage,
                },
              },
            },
          ],
          selectionStrategy: ProductDiscountSelectionStrategy.First,
        },
      });
    }
  }

  // Then add order discounts if available and allowed
  if (hasOrderDiscountClass && orderPercentage > 0) {
    operations.push({
      orderDiscountsAdd: {
        candidates: [
          {
            message: `${orderPercentage}% OFF ORDER`,
            targets: [
              {
                orderSubtotal: {
                  excludedCartLineIds: [],
                },
              },
            ],
            value: {
              percentage: {
                value: orderPercentage,
              },
            },
          },
        ],
        selectionStrategy: OrderDiscountSelectionStrategy.First,
      },
    });
    // [END discount-function.run.cart.add-operations]
  }

  return { operations };
}

// [START discount-function.run.cart.parse-metafield]
function parseMetafield(metafield) {
  try {
    const value = JSON.parse(metafield.value);
    return {
      cartLinePercentage: value.cartLinePercentage || 0,
      orderPercentage: value.orderPercentage || 0,
      collectionIds: value.collectionIds || [],
    };
  } catch (error) {
    console.error("Error parsing metafield", error);
    return {
      cartLinePercentage: 0,
      orderPercentage: 0,
      collectionIds: [],
    };
  }
}
// [END discount-function.run.cart.parse-metafield]
// [END discount-function.run.cart]

File: examples/rust/metafield/extensions/discount-function/src/generate_cart_run.rs
use serde::Deserialize;
use shopify_function::prelude::*;
use shopify_function::Result;

use cart_lines_discounts_generate_run::output::{
    CartLineTarget, CartLinesDiscountsGenerateRunResult, CartOperation, OrderDiscountCandidate,
    OrderDiscountCandidateTarget, OrderDiscountCandidateValue, OrderDiscountSelectionStrategy,
    OrderDiscountsAddOperation, OrderSubtotalTarget, Percentage, ProductDiscountCandidate,
    ProductDiscountCandidateTarget, ProductDiscountCandidateValue,
    ProductDiscountSelectionStrategy, ProductDiscountsAddOperation,
};

use cart_lines_discounts_generate_run::input::{
    InputCartLinesMerchandise::ProductVariant, ResponseData,
};

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct DiscountConfiguration {
    cart_line_percentage: f64,
    order_percentage: f64,
    collection_ids: Vec<String>,
}

// [START discount-function.run.cart]
#[shopify_function_target(
    target = "cartLinesDiscountsGenerateRun",
    query_path = "src/generate_cart_run.graphql",
    schema_path = "schema.graphql"
)]
fn generate_cart_run(input: ResponseData) -> Result<CartLinesDiscountsGenerateRunResult> {
    // [START discount-function.run.cart.parse-metafield]
    let discount_configuration = serde_json::from_str::<DiscountConfiguration>(
        &input
            .discount
            .metafield
            .ok_or("No metafield provided")?
            .value,
    )?;
    // [END discount-function.run.cart.parse-metafield]
    // [START discount-function.run.cart.add-operations]
    let mut operations = vec![];
    if discount_configuration.order_percentage > 0.0 {
        operations.push(CartOperation::OrderDiscountsAdd(
            OrderDiscountsAddOperation {
                selection_strategy: OrderDiscountSelectionStrategy::FIRST,
                candidates: vec![OrderDiscountCandidate {
                    targets: vec![OrderDiscountCandidateTarget::OrderSubtotal(
                        OrderSubtotalTarget {
                            excluded_cart_line_ids: vec![],
                        },
                    )],
                    message: Some(format!(
                        "{}% OFF ORDER",
                        discount_configuration.order_percentage
                    )),
                    value: OrderDiscountCandidateValue::Percentage(Percentage {
                        value: Decimal(discount_configuration.order_percentage),
                    }),
                    conditions: None,
                    associated_discount_code: None,
                }],
            },
        ));
    }

    if discount_configuration.cart_line_percentage > 0.0 {
        let mut cart_line_targets = vec![];
        for line in &input.cart.lines {
            // [START discount-function.run.cart.product.in_any_collection]
            if let ProductVariant(variant) = &line.merchandise {
                if variant.product.in_any_collection
                    || discount_configuration.collection_ids.is_empty()
                {
                    cart_line_targets.push(ProductDiscountCandidateTarget::CartLine(
                        CartLineTarget {
                            id: line.id.clone(),
                            quantity: None,
                        },
                    ));
                }
            }
            // [END discount-function.run.cart.product.in_any_collection]
        }

        if !cart_line_targets.is_empty() {
            operations.push(CartOperation::ProductDiscountsAdd(
                ProductDiscountsAddOperation {
                    selection_strategy: ProductDiscountSelectionStrategy::FIRST,
                    candidates: vec![ProductDiscountCandidate {
                        targets: cart_line_targets,
                        message: Some(format!(
                            "{}% OFF PRODUCT",
                            discount_configuration.cart_line_percentage
                        )),
                        value: ProductDiscountCandidateValue::Percentage(Percentage {
                            value: Decimal(discount_configuration.cart_line_percentage),
                        }),
                        associated_discount_code: None,
                    }],
                },
            ));
        }
    }
    // [END discount-function.run.cart.add-operations]
    Ok(CartLinesDiscountsGenerateRunResult {
        operations: operations,
    })
}
// [END discount_function.run.cart]
